cs/di-extensions.texy

```diff
@@ -85,61 +127,82 @@ services:
 
 /--php
-public function loadConfiguration()
+class BlogExtension extends Nette\DI\CompilerExtension
 {
-	$this->config += $this->defaults;
-	$builder = $this->getContainerBuilder();
-
-	// načtení konfiguračního souboru pro rozšíření
-	$this->compiler->loadDefinitions($builder, $this->loadFromFile(__DIR__ . '/blog.neon')['services'], $this->name);
-
-	// počet článků na stránku v komponentě
-	$builder->getDefinition($this->prefix('articlesList'))
-		->addSetup('setPostsPerPage', [$this->config['postsPerPage']]);
-
-	// volitelné vypnutí komentářů
-	if (!$this->config['comments']) {
-		$builder->getDefinition($this->prefix('comments'))
-			->addSetup('disableComments');
+	public function loadConfiguration()
+	{
+		// načtení konfigurační parametrů
+		$config = $this->validateConfig($this->defaults);
+		$builder = $this->getContainerBuilder();
+
+		// načtení konfiguračního souboru pro rozšíření
+		$this->compiler->loadDefinitions($builder, $this->loadFromFile(__DIR__ . '/blog.neon')['services'], $this->name);
+
+		// počet článků na stránku v komponentě
+		$builder->getDefinition($this->prefix('articlesList'))
+			->addSetup('setPostsPerPage', [$config['postsPerPage']]);
+
+		// volitelné vypnutí komentářů
+		if (!$config['comments']) {
+			$builder->getDefinition($this->prefix('comments'))
+				->addSetup('disableComments');
+		}
 	}
 }
 \--
 
 
-CompilerExtension::beforeCompile() .{toc: beforeCompile()}
-==========================================================
+beforeCompile() .[method]
+=========================
 
-V této fázi sestavování už by neměly přibývat další služby. Můžeme ovšem upravovat existující a doplnit některé potřebné vazby mezi službami, například pomocí tagů.
+V této fázi sestavování už by neměly přibývat další služby. Můžeme ovšem upravovat existující a doplnit některé potřebné vazby mezi službami, například pomocí tagů (`findByTag()`) nebo podle typů (`findByType()`).
 
+/--php
+class BlogExtension extends Nette\DI\CompilerExtension
+{
+	public function beforeCompile()
+	{
+		$builder = $this->getContainerBuilder();
+
+		foreach ($builder->findByTag('logaware')) as $name) {
+			$builder->getDefinition($name)->addSetup('setLogger');
+		}
+	}
+}
+\--
 
-CompilerExtension::afterCompile() .{toc: afterCompile()}
-========================================================
 
-V této fázi už je třída `Container`u vygenerována, obsahuje všechny metody, které vytváří služby a je připravena na zápis do cache. Díky api třídy [Nette\PhpGenerator\ClassType |api:] můžeme přidávat vlastní kód do kontejneru a upravovat tak výsledný kód, který se stará o vytvoření služeb.
+afterCompile() .[method]
+========================
+
+V této fázi už je třída `Container`u vygenerována, obsahuje všechny metody, které vytváří služby a je připravena na zápis do cache. Díky API třídy [Nette\PhpGenerator\ClassType |api:] můžeme přidávat vlastní kód do kontejneru a upravovat tak výsledný kód, který se stará o vytvoření služeb.
```

Comment:
V této fázi už je třída kontejneru vygenerována v podobě objektu [Nette\PhpGenerator\ClassType |api:], obsahuje všechny metody, které vytváří služby, a je připravena na zápis do cache. Výsledný kód třídy můžeme v této chvíli ještě upravit.